In [2]:


import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from tensorflow import keras
import tensorflow as tf
from tensorflow.keras import layers
import math
import pandas as pd
from IPython.display import display

In [6]:

ion = pd.read_csv('../input/dlassign/data1_0.5.csv',header=None)
ion.columns = ["V1", "V2", "Class"]
display(ion.head())

df = ion.copy()


df_train = df.sample(frac=0.7, random_state=0)
df_valid = df.drop(df_train.index)

max_ = df_train.max(axis=0)
min_ = df_train.min(axis=0)

df_train = (df_train - min_) / (max_ - min_)
df_valid = (df_valid - min_) / (max_ - min_)
df_train.dropna(axis=1, inplace=True) 
df_valid.dropna(axis=1, inplace=True)

X_train = df_train.drop('Class', axis=1)
X_valid = df_valid.drop('Class', axis=1)
y_train = df_train['Class']
y_valid = df_valid['Class']

V1        V2  Class
0  0.046000 -0.181767      0
1  0.102303  0.116327      0
2 -0.131546 -0.038680      0
3  0.007224 -0.067146      0
4  0.112290  0.040584      0

In [11]:

activationList = ["relu","sigmoid",'softmax']

for name in activationList:    
    print('Activation :: ',name)
    
    lossList = ["binary_crossentropy","hinge",'squared_hinge']
    
    for lossName in lossList:
        model = keras.Sequential([
        layers.Dense(16, activation=name),
        layers.Dense(4, activation=name),    
        layers.Dense(1, activation='sigmoid'),
        ])  
        print('')
        print('Loss Function Used :: ', lossName)
            
        model.compile(            
            optimizer=tf.keras.optimizers.Adam(),
            loss="binary_crossentropy",
            metrics=['binary_accuracy'],
                )
        early_stopping = keras.callbacks.EarlyStopping(
        patience=10,
        min_delta=0.001,
        restore_best_weights=True,
            )

        history = model.fit(
            X_train, y_train,
            validation_data=(X_valid, y_valid),
            batch_size=512,
            epochs=500,
            callbacks=[early_stopping],            
            verbose=0, # hide the output because we have so many epochs
        )
        history_df = pd.DataFrame(history.history)


        print(("Best Validation Loss: {:0.4f}" +\
              "\nBest Validation Accuracy: {:0.4f}")\
              .format(history_df['val_loss'].min(), 
                      history_df['val_binary_accuracy'].max()))
        print('')
        


Activation ::  relu

Loss Function Used ::  binary_crossentropy
Best Validation Loss: 0.6786
Best Validation Accuracy: 0.5724


Loss Function Used ::  hinge
Best Validation Loss: 0.6834
Best Validation Accuracy: 0.6078


Loss Function Used ::  squared_hinge
Best Validation Loss: 0.6759
Best Validation Accuracy: 0.6219

Activation ::  sigmoid

Loss Function Used ::  binary_crossentropy
Best Validation Loss: 0.6885
Best Validation Accuracy: 0.5724


Loss Function Used ::  hinge
Best Validation Loss: 0.6831
Best Validation Accuracy: 0.5724


Loss Function Used ::  squared_hinge
Best Validation Loss: 0.6855
Best Validation Accuracy: 0.5724

Activation ::  softmax

Loss Function Used ::  binary_crossentropy
Best Validation Loss: 0.6829
Best Validation Accuracy: 0.5724


Loss Function Used ::  hinge
Best Validation Loss: 0.6830
Best Validation Accuracy: 0.5830


Loss Function Used ::  squared_hinge
Best Validation Loss: 0.6824
Best Validation Accuracy: 0.5724




### Based on the above result we conclude the below points

 #### 1) we are considering model with hidden layers having RELU activation and output layer having Sigmoid activation function as it gives maximum validation accuracy and minimum validation loss compared to the other activation functions
 
 #### 2) we are considering "hinge" as a loss function as it giving better result compared to other loss function shown in above result

 #### 3) For learning rate, according to industry experts fixed learnng rate will take more to train the model and it not advisable for real time project so instead of going for fixed learing rate for all the epochs we are using early stopping method which dynamically change the learning rate based on epoch accuracy and loss 



### Maximum Accuracy Achieved : 0.6219
### Minimum Loss Achieved : 0.6759